watertap/property_models/anaerobic_digestion/adm1_reactions.py

```diff
@@ -0,0 +1,1734 @@
+                == self.conc_mass_comp_ref["S_IN"]
+                / (14 * pyo.units.kg / pyo.units.kmol)
+                - self.conc_mol_nh3
+            )
+
+        self.concentration_of_nh4 = pyo.Constraint(
+            rule=concentration_of_nh4_rule,
+            doc="constraint concentration of pro-",
+        )
+
+        def S_OH_rule(self):
+            return self.S_OH == self.params.KW / self.S_H
+
+        self.S_OH_cons = pyo.Constraint(
+            rule=S_OH_rule,
+            doc="constraint concentration of OH",
+        )
+
+        def S_H_rule(self):
+            return (
+                self.state_ref.cations
+                + self.conc_mol_nh4
+                + self.S_H
+                - self.conc_mol_hco3
+                - self.conc_mass_ac / (64 * pyo.units.kg / pyo.units.kmol)
+                - self.conc_mass_pro / (112 * pyo.units.kg / pyo.units.kmol)
+                - self.conc_mass_bu / (160 * pyo.units.kg / pyo.units.kmol)
+                - self.conc_mass_va / (208 * pyo.units.kg / pyo.units.kmol)
+                - self.S_OH
+                - self.state_ref.anions
+                == 0
+            )
+
+        self.S_H_cons = pyo.Constraint(
+            rule=S_H_rule,
+            doc="constraint concentration of pro-",
+        )
+
+        def rule_pH(self):
+            return -pyo.log10(self.S_H / pyo.units.kmol * pyo.units.m**3)
+
+        self.pH = pyo.Expression(rule=rule_pH, doc="pH of solution")
+
+        def rule_I_IN_lim(self):
+            return 1 / (
+                1 + self.params.K_S_IN / (self.conc_mass_comp_ref["S_IN"] / mw_n)
+            )
+
+        self.I_IN_lim = pyo.Expression(
+            rule=rule_I_IN_lim,
+            doc="Inhibition function related to secondary substrate; inhibit uptake when inorganic nitrogen S_IN~ 0",
+        )
+
+        def rule_I_h2_fa(self):
+            return 1 / (1 + self.conc_mass_comp_ref["S_h2"] / self.params.K_I_h2_fa)
+
+        self.I_h2_fa = pyo.Expression(
+            rule=rule_I_h2_fa,
+            doc="hydrogen inhibition attributed to long chain fatty acids",
+        )
+
+        def rule_I_h2_c4(self):
+            return 1 / (1 + self.conc_mass_comp_ref["S_h2"] / self.params.K_I_h2_c4)
+
+        self.I_h2_c4 = pyo.Expression(
+            rule=rule_I_h2_c4,
+            doc="hydrogen inhibition attributed to valerate and butyrate uptake",
+        )
+
+        def rule_I_h2_pro(self):
+            return 1 / (1 + self.conc_mass_comp_ref["S_h2"] / self.params.K_I_h2_pro)
+
+        self.I_h2_pro = pyo.Expression(
+            rule=rule_I_h2_pro,
+            doc="hydrogen inhibition attributed to propionate uptake",
+        )
+
+        def rule_I_nh3(self):
+            return 1 / (1 + self.conc_mol_nh3 / self.params.K_I_nh3)
+
+        self.I_nh3 = pyo.Expression(
+            rule=rule_I_nh3, doc="ammonia inibition attributed to acetate uptake"
+        )
+
+        def rule_I_pH_aa(self):
+            return pyo.Expr_if(
```

Comment:
I was concerned about adding these in for potential stability reasons. We may want to push the limits on these implemented expression-ifs to further probe stability.  Tagging #909